plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/sampler/ProbabilisticSampler.java

```diff
@@ -37,27 +46,62 @@ public ProbabilisticSampler(TelemetrySettings telemetrySettings) {
         this.telemetrySettings = Objects.requireNonNull(telemetrySettings);
         this.samplingRatio = telemetrySettings.getSamplingProbability();
         this.defaultSampler = Sampler.traceIdRatioBased(samplingRatio);
+        this.actionBasedSampler = new HashMap<>();
+        this.actionBasedSamplingProbability = new HashMap<>();
     }
 
-    Sampler getSampler() {
+    /**
+     * Returns custom sampler based on the type of request
+     * @param attributes Telemetry attributes
+     */
+    Sampler getSampler(Attributes attributes) {
         double newSamplingRatio = telemetrySettings.getSamplingProbability();
         if (isSamplingRatioChanged(newSamplingRatio)) {
             synchronized (this) {
                 this.samplingRatio = newSamplingRatio;
                 defaultSampler = Sampler.traceIdRatioBased(samplingRatio);
             }
         }
+
+        final String action = attributes.get(AttributeKey.stringKey(TRANSPORT_ACTION));
+        if (action != null && telemetrySettings.isActionSamplingOverrideSet(action)) {
+            if (isActionSamplerUpdateRequired(action)) {
+                synchronized (this) {
+                    updateActionSampler(action);
+                }
+            }
+            return actionBasedSampler.get(action);
+        }
+
         return defaultSampler;
     }
 
     private boolean isSamplingRatioChanged(double newSamplingRatio) {
         return Double.compare(this.samplingRatio, newSamplingRatio) != 0;
     }
 
+    private boolean isActionSamplerUpdateRequired(String action) {
+        return (!actionBasedSampler.containsKey(action)
+            || actionBasedSamplingProbability.get(action) != telemetrySettings.getActionSamplingProbability(action));
+    }
+
+    private void updateActionSampler(String action) {
+        double samplingRatio = telemetrySettings.getActionSamplingProbability(action);
+        this.actionBasedSamplingProbability.put(action, samplingRatio);
```

Comment:
Can you please keep this in a single Map?

Reply:
Since only few override will be present at a action level, so I think it's fine to keep 2 maps for better readability. Please do let me know if that's fine.